pkg/ui/workspaces/cluster-ui/src/api/databases/grantsApi.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+import useSWRImmutable from "swr/immutable";
+
+import { fetchDataJSON } from "../fetchData";
+import {
+  APIV2ResponseWithPaginationState,
+  SimplePaginationState,
+} from "../types";
+
+export type DatabaseGrant = {
+  grantee: string;
+  privilege: string;
+};
+
+export enum GrantsSortOptions {
+  GRANTEE = "grantee",
+  PRIVILEGE = "privilege",
+}
+
+type DatabaseGrantsRequest = {
+  dbId: number;
+  sortBy?: GrantsSortOptions;
+  sortOrder?: "asc" | "desc";
+  pagination?: SimplePaginationState;
+};
+
+export type DatabaseGrantsResponse = APIV2ResponseWithPaginationState<
+  DatabaseGrant[]
+>;
+
+const createDbGrantsPath = (req: DatabaseGrantsRequest): string => {
+  const { dbId, pagination, sortBy, sortOrder } = req;
+  const urlParams = new URLSearchParams();
+  if (pagination?.pageNum) {
+    urlParams.append("pageNum", pagination.pageNum.toString());
+  }
+  if (pagination?.pageSize) {
+    urlParams.append("pageSize", pagination.pageSize.toString());
+  }
+  if (sortBy) {
+    urlParams.append("sortBy", sortBy);
+  }
+  if (sortOrder) {
+    urlParams.append("sortOrder", sortOrder);
+  }
+  return `api/v2/grants/databases/${dbId}/?` + urlParams.toString();
+};
+
+const fetchDbGrants = (
+  req: DatabaseGrantsRequest,
+): Promise<DatabaseGrantsResponse> => {
+  const path = createDbGrantsPath(req);
+  return fetchDataJSON(path);
+};
+
+export const useDatabaseGrants = (req: DatabaseGrantsRequest) => {
```

Comment:
Is this hook used anywhere other than the one page? If so, I'd prefer it to just be in the component for clarity. It's an interesting question whether the use of SWR should be obvious in the component or hidden in a hook.

Also the other thread made me think that if `useMemo` and `SWRImmutable` are connected in some way, it's nicer to have them near each other in the code.

Happy to hear thoughts to the contrary, I know you're trying to organize the "API" stuff in this file. Do we really need a custom hook for every API call?

Reply:
> Also the other thread made me think that if useMemo and SWRImmutable are connected in some way, it's nicer to have them near each other in the code.

I think that's a good call out on the useMemo being useful since it's immutable -- if anything I'd change the naming on the grants hook to `useGatabaseGrantsImmutable` to communicate that over swapping to useSWR directly.

> Do we really need a custom hook for every API call?
For get calls, I think it's nice to standardize on using SWR going forward which manages the req state and client side caching for us. It's true that this is the only usage right now but wrapping them in their own exported hook is also a pretty small lift. Even on cloud, I've seen useSWR being wrapped if there's only a single usage. I kind of like that get requests would be standarized like this so I'd prefer to keep it as is Iwith the immutable rename). Thoughts?